src/codecs/openexr.rs

```diff
@@ -0,0 +1,433 @@
+        // therefore we currently read all lines at once
+        // Todo: optimize for specific exr.line_order?
+       self.total_bytes()
+    }
+
+    // reads with or without alpha, depending on `self.alpha_preference` and `self.alpha_present_in_file`
+    fn read_image_with_progress<F: Fn(Progress)>(self, target: &mut [u8], progress_callback: F) -> ImageResult<()> {
+        let blocks_in_header = self.selected_exr_header().chunk_count as u64;
+        let channel_count = self.color_type().channel_count() as usize;
+
+        let result = read()
+            .no_deep_data().largest_resolution_level()
+            .rgba_channels(
+                |size, _channels| {
+                    // TODO debug_assert_eq!(self.header_index, header_index);
+                    (size, vec![0_f32; size.area() * channel_count])
+                },
+
+                move |(size, buffer), position, (r,g,b,a_or_1): (f32,f32,f32,f32)| {
+                    let first_f32_index = position.flat_index_for_size(*size);
+
+                    buffer[first_f32_index * channel_count .. (first_f32_index + 1) * channel_count]
+                        .copy_from_slice(&[r, g, b, a_or_1][0 .. channel_count]);
+
+                    // TODO white point chromaticities + srgb/linear conversion?
+                }
+            )
+            .first_valid_layer() // TODO select exact layer by self.header_index?
+            .all_attributes()
+            .on_progress(|progress| {
+                progress_callback(Progress::new(
+                    (progress*blocks_in_header as f64) as u64, blocks_in_header) // TODO precision errors?
+                );
+            })
+            .from_chunks(self.exr_reader).map_err(to_image_err)?;
+
+        // TODO this copy is strictly not necessary, but the exr api is a little too simple for reading into a borrowed target slice
+        target.copy_from_slice(bytemuck::cast_slice(result.layer_data.channel_data.pixels.1.as_slice()));
+        Ok(())
+    }
+}
+
+
+
+
+/// Write a raw byte buffer of pixels,
+/// panicking if the buffer is not aligned to `f32`
+/// or if it has an invalid length.
+///
+/// Assumes the writer is buffered. In most cases,
+/// you should wrap your writer in a `BufWriter` for best performance.
+pub fn write_buffer(
+    mut buffered_write: impl Write/* + Seek*/, bytes: &[u8],
+    width: u32, height: u32, color_type: ColorType
+) -> ImageResult<()>
+{
+    let width = width as usize;
+    let height = height as usize;
+
+    let mut seekable_write = Cursor::new(Vec::with_capacity(width*height*3*4)); // TODO remove
+
+    match color_type {
+        ColorType::Rgb32F => {
+            let pixels: &[f32] = bytemuck::try_cast_slice(bytes)
+                .expect("image byte buffer must be aligned to f32");
+
+            debug_assert_eq!(pixels.len(), width * height * 3, "invalid dimensions for byte buffer length");
+
+            exr::prelude::Image
```

Comment:
OpenEXR supports many different compression methods, each with a different a trade-off between speed and compression ratio. By default, a fast but rather weak compression is used (plain run-length encoding). This could also use deflate compression, which is slower, but yields small images. What do you think?